File: fastapi_app/levels.py
"""A level is a unanimous group of reviewers. Forward edges are a derived view."""

from __future__ import annotations


def normalize_levels(graph: dict) -> list[dict]:
    if graph.get("levels") is not None:
        # GraphModel persists the derived forward nodes and edges alongside the
        # canonical levels. Levels remain the source of truth on a round-trip.
        return graph["levels"]
    nodes, edges = graph.get("nodes", []), graph.get("edges", [])
    if not nodes:
        return []
    if any(
        n["node_type"] not in {"start", "reviewer", "join_all", "end"} for n in nodes
    ):
        raise ValueError(
            "Use ordered unanimous review levels; conditional and join-any nodes are unsupported."
        )
    node_keys = [n["node_key"] for n in nodes]
    if len(set(node_keys)) != len(node_keys):
        raise ValueError("The imported workflow contains duplicate node IDs.")
    if (
        sum(n["node_type"] == "start" for n in nodes) != 1
        or sum(n["node_type"] == "end" for n in nodes) != 1
    ):
        raise ValueError(
            "The imported workflow must contain exactly one start and one end node."
        )
    if any(
        e["from_node_key"] not in node_keys or e["to_node_key"] not in node_keys
        for e in edges
    ):
        raise ValueError("The imported workflow contains an edge to an unknown node.")
    if any(
        e.get("condition_json") or e.get("action") not in {None, "always", "approve"}
        for e in edges
    ):
        raise ValueError(
            "The imported workflow contains an unsupported route or condition."
        )
    start = next(n["node_key"] for n in nodes if n["node_type"] == "start")
    end = next(n["node_key"] for n in nodes if n["node_type"] == "end")
    if any(e["to_node_key"] == start for e in edges) or any(
        e["from_node_key"] == end for e in edges
    ):
        raise ValueError("The imported workflow has an invalid start or end route.")
    if any(e["from_node_key"] == e["to_node_key"] for e in edges):
        raise ValueError("The imported workflow contains a self-referential route.")
    node_types = {node["node_key"]: node["node_type"] for node in nodes}
    for edge in edges:
        source, target = (
            node_types[edge["from_node_key"]],
            node_types[edge["to_node_key"]],
        )
        if source == "start" and (
            edge.get("action") not in {None, "always"}
            or target not in {"reviewer", "join_all"}
        ):
            raise ValueError("The imported workflow has an unsupported start route.")
        if source == "reviewer" and (
            edge.get("action") not in {None, "approve"}
            or target not in {"reviewer", "join_all", "end"}
        ):
            raise ValueError("The imported workflow has an unsupported reviewer route.")
        if source == "join_all" and (
            edge.get("action") not in {None, "always"}
            or target not in {"reviewer", "end"}
        ):
            raise ValueError("The imported workflow has an unsupported join route.")
    pending = set(node_keys)
    depth = {}
    while pending:
        progressed = False
        for node in nodes:
            key = node["node_key"]
            if key not in pending:
                continue
            parents = [e["from_node_key"] for e in edges if e["to_node_key"] == key]
            if key != start and not parents:
                raise ValueError(
                    f'The imported node "{key}" is disconnected from the start.'
                )
            if any(p not in depth for p in parents):
                continue
            depth[key] = max([depth[p] for p in parents] or [0]) + (
                node["node_type"] == "reviewer"
            )
            pending.remove(key)
            progressed = True
        if not progressed:
            raise ValueError("Forward workflow contains a cycle or an unknown node.")
    if not any(e["to_node_key"] == end for e in edges):
        raise ValueError("The imported workflow does not route to its end node.")
    reachable = {start}
    while True:
        expanded = reachable | {
            e["to_node_key"] for e in edges if e["from_node_key"] in reachable
        }
        if expanded == reachable:
            break
        reachable = expanded
    if len(reachable) != len(node_keys):
        raise ValueError(
            "The imported workflow contains nodes unreachable from the start."
        )
    reverse_reachable = {end}
    while True:
        expanded = reverse_reachable | {
            e["from_node_key"] for e in edges if e["to_node_key"] in reverse_reachable
        }
        if expanded == reverse_reachable:
            break
        reverse_reachable = expanded
    if len(reverse_reachable) != len(node_keys):
        raise ValueError(
            "The imported workflow contains nodes that cannot reach the end."
        )
    groups = {}
    for node in nodes:
        if node["node_type"] != "reviewer":
            continue
        d = depth[node["node_key"]]
        meta = node.get("metadata") or {}
        group = groups.setdefault(
            d,
            {
                "id": meta.get("level_id", f"level_{d}"),
                "name": meta.get("level_name", f"Level {d}"),
                "reviewers": [],
            },
        )
        group["reviewers"].append(node)
    return [groups[key] for key in sorted(groups)]


def compile_levels(levels: list[dict]) -> tuple[list[dict], list[dict]]:
    def connector(key, kind):
        return {
            "node_key": key,
            "node_type": kind,
            "visible_sections": [],
            "allowed_actions": [],
            "metadata": {"required_inputs": []},
        }

    level_ids = [level["id"] for level in levels]
    reviewer_ids = [
        reviewer["node_key"] for level in levels for reviewer in level["reviewers"]
    ]
    reserved_ids = {"start", "end", *(f"join_{level_id}" for level_id in level_ids)}
    if set(reviewer_ids) & reserved_ids:
        raise ValueError("A reviewer ID conflicts with a generated connector ID.")
    nodes = [connector("start", "start")]
    edges = []
    previous = "start"
    for level in levels:
        join = f"join_{level['id']}"
        for reviewer in level["reviewers"]:
            node = {
                **reviewer,
                "metadata": {
                    **reviewer.get("metadata", {}),
                    "level_id": level["id"],
                    "level_name": level["name"],
                },
            }
            nodes.append(node)
            edges.extend(
                [
                    {
                        "from_node_key": previous,
                        "to_node_key": node["node_key"],
                        "action": "always",
                    },
                    {
                        "from_node_key": node["node_key"],
                        "to_node_key": join,
                        "action": "approve",
                    },
                ]
            )
        nodes.append(connector(join, "join_all"))
        previous = join
    nodes.append(connector("end", "end"))
    edges.append({"from_node_key": previous, "to_node_key": "end", "action": "always"})
    return nodes, edges

File: src/components/admin/levels.ts
import type { StudioGraphNode, StudioGraphEdge } from "./studioTypes";

export type ReviewLevel = { id: string; name: string; reviewers: StudioGraphNode[] };
export const id = (prefix: string) => `${prefix}_${crypto.randomUUID().replaceAll("-", "")}`;
export function reviewer(): StudioGraphNode {
  return { node_key: id("review"), node_type: "reviewer", display_name: "New reviewer", reviewer_email: "",
    visible_sections: [], allowed_actions: ["approve", "request_changes", "comment"],
    metadata: { required_inputs: [], sla_hours: 72, can_view_comments: false, return_target: "student", return_rule: null, student_visible_fields: [] } };
}
export function newLevel(): ReviewLevel {
  return { id: id("level"), name: "Review level", reviewers: [reviewer()] };
}

/** Import the supported old start/reviewer/join_all/end format once at the boundary. */
export function levelsFromGraph(nodes: StudioGraphNode[], edges: StudioGraphEdge[]): ReviewLevel[] {
  if (!nodes.length) return [];
  if (nodes.some((n) => !["start", "reviewer", "join_all", "end"].includes(n.node_type))) {
    throw new Error("This draft uses conditional or any-reviewer branches. Rebuild it as unanimous review levels.");
  }
  const keys = nodes.map((node) => node.node_key);
  if (new Set(keys).size !== keys.length) throw new Error("The imported workflow contains duplicate node IDs.");
  if (nodes.filter((node) => node.node_type === "start").length !== 1 || nodes.filter((node) => node.node_type === "end").length !== 1) {
    throw new Error("The imported workflow must contain exactly one start and one end node.");
  }
  if (edges.some((edge) => !keys.includes(edge.from_node_key) || !keys.includes(edge.to_node_key))) {
    throw new Error("The imported workflow contains an edge to an unknown node.");
  }
  const pending = new Set(keys);
  const depth = new Map<string, number>();
  while (pending.size) {
    let progressed = false;
    for (const node of nodes.filter((n) => pending.has(n.node_key))) {
      const parents = edges.filter((e) => e.to_node_key === node.node_key && !["request_changes", "reject"].includes(e.action || ""));
      if (node.node_type !== "start" && !parents.length) throw new Error(`The imported node "${node.node_key}" is disconnected from the start.`);
      if (parents.some((e) => !depth.has(e.from_node_key))) continue;
      depth.set(node.node_key, Math.max(0, ...parents.map((e) => depth.get(e.from_node_key)!)) + (node.node_type === "reviewer" ? 1 : 0));
      pending.delete(node.node_key); progressed = true;
    }
    if (!progressed) throw new Error("The forward review flow contains a cycle or an unknown node.");
  }
  const grouped = new Map<number, ReviewLevel>();
  for (const node of nodes.filter((n) => n.node_type === "reviewer")) {
    const d = depth.get(node.node_key)!;
    const levelId = String(node.metadata.level_id || `level_${d}`);
    const levelName = String(node.metadata.level_name || `Level ${d}`);
    const existing = grouped.get(d);
    if (existing && (existing.id !== levelId || existing.name !== levelName)) {
      throw new Error(`Reviewers at depth ${d} disagree about their level identity.`);
    }
    const level = existing || { id: levelId, name: levelName, reviewers: [] };
    level.reviewers.push(node); grouped.set(d, level);
  }
  return validateLevels([...grouped.entries()].sort(([a], [b]) => a - b).map(([, level]) => level));
}

export function validateLevels(levels: ReviewLevel[]): ReviewLevel[] {
  const validId = /^[A-Za-z0-9_-]+$/;
  const levelIds = levels.map((level) => level.id);
  const nodeIds = levels.flatMap((level) => level.reviewers.map((node) => node.node_key));
  if (levelIds.some((levelId) => !validId.test(levelId))) throw new Error("The imported workflow contains an invalid level ID.");
  if (new Set(levelIds).size !== levelIds.length) throw new Error("The imported workflow contains duplicate level IDs.");
  if (nodeIds.some((nodeId) => !validId.test(nodeId))) throw new Error("The imported workflow contains an invalid reviewer ID.");
  if (new Set(nodeIds).size !== nodeIds.length) throw new Error("The imported workflow contains duplicate reviewer IDs.");
  const connectorIds = new Set(["start", "end", ...levelIds.map((levelId) => `join_${levelId}`)]);
  if (nodeIds.some((nodeId) => connectorIds.has(nodeId))) throw new Error("A reviewer ID conflicts with a generated connector ID.");
  if (levels.some((level) => level.reviewers.some((node) => node.node_type !== "reviewer"))) {
    throw new Error("Review levels may contain only reviewer nodes.");
  }
  return levels;
}

/** Forward edges are derived. They are never independently editable. */
export function compileLevels(levels: ReviewLevel[]) {
  validateLevels(levels);
  const nodes: StudioGraphNode[] = [{ node_key: "start", node_type: "start", visible_sections: [], allowed_actions: [], metadata: { required_inputs: [] } }];
  const edges: StudioGraphEdge[] = [];
  let previous = "start";
  for (const level of levels) {
    const join = `join_${level.id}`;
    nodes.push(...level.reviewers.map((n) => ({ ...n, metadata: { ...n.metadata, level_id: level.id, level_name: level.name } })));
    for (const n of level.reviewers) {
      edges.push({ from_node_key: previous, to_node_key: n.node_key, action: "always" });
      edges.push({ from_node_key: n.node_key, to_node_key: join, action: "approve" });
    }
    nodes.push({ node_key: join, node_type: "join_all", visible_sections: [], allowed_actions: [], metadata: { required_inputs: [] } });
    previous = join;
  }
  nodes.push({ node_key: "end", node_type: "end", visible_sections: [], allowed_actions: [], metadata: { required_inputs: [] } });
  edges.push({ from_node_key: previous, to_node_key: "end", action: "always" });
  return { levels, nodes, edges };
}
